components/mediators/cache/org.wso2.carbon.mediator.cache/src/main/java/org/wso2/carbon/mediator/cache/CacheMediator.java

```diff
@@ -0,0 +1,693 @@
+			synLog.traceOrDebug("Looking up cache at scope : " + scope + " with ID : " + cacheKey);
+		}
+
+		boolean result = true;
+		try {
+			if (synCtx.isResponse()) {
+				processResponseMessage(synCtx, cfgCtx, synLog);
+
+			} else {
+				result = processRequestMessage(synCtx, synLog);
+			}
+
+		} catch (ClusteringFault clusteringFault) {
+			synLog.traceOrDebug("Unable to replicate Cache mediator state among the cluster");
+		}
+
+		synLog.traceOrDebug("End : Cache mediator");
+
+		return result;
+	}
+
+	/**
+	 * Process a response message through this cache mediator. This finds the Cache used, and
+	 * updates it for the corresponding request hash
+	 *
+	 * @param synLog the Synapse log to use
+	 * @param synCtx the current message (response)
+	 * @param cfgCtx the abstract context in which the cache will be kept
+	 * @throws ClusteringFault is there is an error in replicating the cfgCtx
+	 */
+	private void processResponseMessage(MessageContext synCtx, ConfigurationContext cfgCtx,
+	                                    SynapseLog synLog) throws ClusteringFault {
+
+		if (!collector) {
+			handleException("Response messages cannot be handled in a non collector cache", synCtx);
+		}
+		org.apache.axis2.context.MessageContext msgCtx = ((Axis2MessageContext) synCtx).getAxis2MessageContext();
+		OperationContext operationContext = msgCtx.getOperationContext();
+		CachableResponse response = (CachableResponse) operationContext.getProperty(CachingConstants.CACHED_OBJECT);
+
+		if (response != null) {
+			if (synLog.isTraceOrDebugEnabled()) {
+				synLog.traceOrDebug("Storing the response message into the cache at scope : " + scope + " with ID : "
+				                    + cacheKey + " for request hash : " + response.getRequestHash());
+			}
+			if (synLog.isTraceOrDebugEnabled()) {
+				synLog.traceOrDebug("Storing the response for the message with ID : " + synCtx.getMessageID() + " " +
+				                    "with request hash ID : " + response.getRequestHash() + " in the cache : " +
+				                    cacheKey);
+			}
+
+			ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+			try {
+				synCtx.getEnvelope().serialize(outStream);
+				response.setResponseEnvelope(outStream.toByteArray());
+				if (msgCtx.isDoingREST()) {
+					response.setSOAP11(synCtx.isSOAP11());
+					Map<String, String> headers =
+							(Map) msgCtx.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+					String messageType = (String) msgCtx.getProperty(Constants.Configuration.MESSAGE_TYPE);
+					Map<String, Object> headerProperties = new HashMap<String, Object>();
+					headerProperties.put(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS, headers);
+					headerProperties.put(Constants.Configuration.MESSAGE_TYPE, messageType);
+					response.setHeaderProperties(headerProperties);
+				}
+
+			} catch (XMLStreamException e) {
+				handleException("Unable to set the response to the Cache", e, synCtx);
+			}
```

Comment:
Yes, We should close outStream in finally block. Thanks a lot for pointing this out. Issue is fixed with PR #452.